javascript/ql/test/query-tests/Security/CWE-352/MissingCsrfMiddlewareGood2.js

```diff
@@ -0,0 +1,75 @@
+var express = require('express');
+var cookieParser = require('cookie-parser');
+var passport = require('passport');
+
+(function () {
+
+    var app = express()
+
+    app.use(cookieParser())
+    app.use(passport.authorize({ session: true }))
+
+    function getCsrfToken(request) {
+        return request.headers['x-xsrf-token'];
+    }
+
+    function setCsrfToken(request, response, next) {
+        response.cookie('XSRF-TOKEN', request.csrfToken());
+        next();
+    }
+
+    const csrf = {
+        getCsrfToken: getCsrfToken,
+        setCsrfToken: setCsrfToken
+    };
+
+    app.use(express.csrf({ value: csrf.getCsrfToken }));
```

Comment:
What is `express.csrf`? Does that offer CSRF protection? Because it looks like the test has two CSRF middlewares now.

Reply:
It does half of the CSRF protection.    
`express.csrf` is responsible for checking the CSRF token. 
And `csrf.setCsrfToken` is responsible for setting the CSRF cookie.   

`express.csrf` has been deprecated in favor of `csurf`, so you will only see this pattern in older codebases.   
(Altough it is easy [to find uses on github](https://github.com/search?q=app.use%28express.csrf+extension%3Ajs&ref=simplesearch)).

Reply:
Shouldn't we add `express.csrf` to `csrfMiddlewareCreation` then? If it's the one that performs the check, then it's the half that matters for security.